rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -0,0 +1,294 @@
+package coop.rchain.rholang.interpreter
+
+import coop.rchain.rholang.interpreter.Env.DeBruijn
+
+object Substitute {
+
+  def subOrDec(term: Var)(implicit env: Env[Par]): Either[Var, Par] =
+    term match {
+      case BoundVar(level) =>
+        env.get(level) match {
+          case Some(par) => Right(rename(par, env.level))
+          case None =>
+            if (level - env.level < 0)
+              throw new Error(s"Illegal Index [Variable: $term, Level: ${level - env.level}]")
+            else Left(BoundVar(level - env.level))
+        }
+      case _ => throw new Error(s"Illegal Substitution [$term]")
+    }
+
+  def subOrDec(exp: EVar)(implicit env: Env[Par]): Either[EVar, Par] =
+    subOrDec(exp.v) match {
+      case Left(varue) => Left(EVar(varue))
+      case Right(par)  => Right(par)
+    }
+
+  def substitute(term: Channel)(implicit env: Env[Par]): Channel =
+    term match {
+      case Quote(p) => Quote(substitute(p))
+      case ChanVar(v) =>
+        subOrDec(v) match {
+          case Left(_v) => ChanVar(_v)
+          case Right(p) => Quote(p)
+        }
+    }
+
+  def substitute(term: Par)(implicit env: Env[Par]): Par = {
+
+    def subExp(expxs: List[Expr]): Par =
+      (Par() /: expxs) { (par, expr) =>
+        expr match {
+          case e @ EVar(_) =>
+            subOrDec(e) match {
+              case Left(_e)    => par :+ _e
+              case Right(_par) => par ++ _par
+            }
+          case e @ _ => par :+ substitute(e)
```

Comment:
It's probably better to go through the list in reverse and prepend to the result list.